### ISSUE TITLE
Document how to add/remove glyphs in production Google Sans source files

### ISSUE THREAD

**Issue by chrissimpkins** [googlefonts/googlesans#369]
There have been a number of recent changes to the import process and production file normalization workflow. I tried to add a new baht.tf glyph definition to our prod sources as a composite based on the updated baht outlines coming in with #365.  This exercise should take 5 mins. I spent multiple hours on it.  

It is clear that we do not support direct Robofont writes to our production files, even followed by a R/W round trip through ufoLib2 and our normalizer script. I used a separate set of production files and attempted to create a new glyph with ufoLib2, then update it with the new outline and this process does not appear to write the required lib.plist data for the glyph addition.  We do not want to import all lib data coming out of a Robofont write.  I ended up resetting all of the Robofont lib.plist files and hand editing the lib.plist in every master with the new glyph data in the glyphOrder and postscript names fields.  It looks like baht does not require the GDEF definitions that are now included in the UFO files but a new glyph that does will require edits there too.  

Let's (1) document how to add (and remove) a glyph in prod files now; (2) review the full process based on this documentation to confirm that we have not become so fancy that easy things are no longer easy.  We want to be able to support font editor use on our production sources.

cc @madig @belluzj

**Comment by chrissimpkins** [comment]
Robofont v4 lib.plist write diff

[diff.diff.zip](https://github.com/googlefonts/googlesans/files/7048441/diff.diff.zip)

@madig

**Comment by madig** [comment]
Robofont seems to reorder the glyphs? Not sure how to make it stop doing that.

**Comment by chrissimpkins** [comment]
It edits fontinfo.plist and metainfo.plist too.  It is not ideal, but not a significant problem to write out a separate set of UFO sources and import into the prod sources with ufoLib2/another serializer lib.  But we don't appear to have a one step solution to address additions/removals with that approach.  You may have a better solution.  Let's document how you would approach it and then we can decide if there are any other actions to take here.  I didn't recognize that this was an issue until I ran an editor over the production source files for a glyph addition.

**Comment by madig** [comment]
Ok. So, I do the following with .glyphs sources (`import.txt` is a one-per-line glyph list):

```
$ python scripts/gs-glyphs2ufo.py source/GoogleSans/staging/*.glyphs
$ python scripts/gs-merge-designspace.py --source source/GoogleSans/staging/GoogleSansTHLooped.designspace --target source/GoogleSans/GoogleSans.designspace --import-glyphs-file import.txt
$ python scripts/gs-merge-designspace.py --source source/GoogleSans/staging/GoogleSansTHLooped-Italic.designspace --target source/GoogleSans/GoogleSans-Italic.designspace --import-glyphs-file import.txt
$ python scripts/gs-normalize-designspace.py  # For good measure
```

(The Designspaces may need editing to match the upstream axis setup.)

After importing features by hand, I run

```
$ python scripts/format-fea.py source/GoogleSans/thai.fea
```

And format some stuff by hand.

Then one can look at the main sources lib.plist and check the additions to `public.postscriptNames` and `public.openTypeCategories`.

**Comment by chrissimpkins** [comment]
It looks we can skip the first stage and start at the gs-merge-designspace.py step with the directly edited UFO files?

**Comment by chrissimpkins** [comment]
> It looks we can skip the first stage and start at the gs-merge-designspace.py step with the directly edited UFO files?

This worked! 👍 

#371

**Comment by madig** [comment]
Yep.

**Comment by chrissimpkins** [comment]
@madig @belluzj Let's use https://github.com/googlefonts/googlesans/issues/380 as a test case to document the full approach to remove a glyph from the glyph set.

**Comment by belluzj** [comment]
I opened #382 to which the documentation about that process can be added by Nikolaus next week

**Comment by chrissimpkins** [comment]
Example glyph removal source from https://github.com/googlefonts/googlesans/pull/379#issuecomment-915342026:

```python
from pathlib import Path

from ufoLib2 import Font

for p in Path("source/GoogleSans/").glob("*.ufo"):
    ufo = Font.open(p)
    for layer in ufo.layers:
        if "Eng-tail" in layer:
            del layer["Eng-tail"]
    if "Eng-tail" in ufo.lib["public.glyphOrder"]:
        ufo.lib["public.glyphOrder"].remove("Eng-tail")
    ufo.save()
```

**Comment by chrissimpkins** [comment]
> Example glyph removal source from [#379 (comment)](https://github.com/googlefonts/googlesans/pull/379#issuecomment-915342026):
> 
> ```python
> from pathlib import Path
> 
> from ufoLib2 import Font
> 
> for p in Path("source/GoogleSans/").glob("*.ufo"):
>     ufo = Font.open(p)
>     for layer in ufo.layers:
>         if "Eng-tail" in layer:
>             del layer["Eng-tail"]
>     if "Eng-tail" in ufo.lib["public.glyphOrder"]:
>         ufo.lib["public.glyphOrder"].remove("Eng-tail")
>     ufo.save()
> ```

Manual glyph name removal is required for the GDEF lists correct?  Is the write to GDEF lists automated on import in some fashion?

**Comment by madig** [comment]
Yes there are more places that need to be looked at, but I think the normalization script will catch a few. It was enough for Eng-tail. An easy test is searching for the glyph name you want to remove and then expand the script to clear out all places you find.

GDEF definitions are updated on normalization.

**Comment by chrissimpkins** [comment]
> An easy test is searching for the glyph name you want to remove and then expand the script to clear out all places you find.

Let's assume that at some stage in the life cycle of this project none of us will be involved and this should be evident to whoever is maintaining it in the docs.  Ideally, it would be nice to support `python add-glyph.py [GLYPHNAME]` and `python drop-glyph.py [GLYPHNAME]` to wrap all of the necessary steps into a one-liner in the future.  Then we can drop the documentation of these issues and point to the scripts.

**Comment by madig** [comment]
Ok, adding glyphs is currently done through merging a Designspace in. I can make a quick script for dropping it when I get to it.

**Comment by MariannaPaszkowska** [comment]
note to self: check if this has been documented

**Comment by MariannaPaszkowska** [comment]
@madig We should make sure we update it with the information about our separate families, as we update our cross-programme build process.
